conftest.py

```diff
@@ -453,6 +281,280 @@ def tracy_profile():
+        parent_pid = os.getpid()
+        logger.info(f"Metal timeout {settings.timeout} seconds {parent_pid} for {item.nodeid}")
+
+        def get_parent_status():
+            try:
+                parent = psutil.Process(parent_pid)
+            except:
+                return "already dead"
+            return parent.status()
+
+        def run_timer(settings):
+            logger.info(f"Timer started for {item.nodeid}")
+            dead_status = ["zombie", "dead", "already dead"]
+            timeout = settings.timeout
+            parent_status = "running"
+            while parent_status not in dead_status and timeout > 0:
+                time.sleep(5)
+                timeout -= 5
+                parent_status = get_parent_status()
+            if parent_status != "already dead":
+                logger.info(f"Timing out test case")
+                os.kill(parent_pid, signal.SIGKILL)
+            logger.info(f"Killing timer")
+            os._exit(1)
+
+        def cancel():
+            logger.info(f"Cancelling timer")
+            metal_timer.terminate()
+
+        metal_timer = multiprocess.Process(target=run_timer, args=(settings,), daemon=True)
+        item.cancel_timeout = cancel
+        metal_timer.start()
+        # logger.info(f"parent and metal timer pid: {parent_pid} {metal_timer.pid}")
+    return True
+
+
+# This is a hook used in pytest-xdist to handle when a worker crashes out
+# In our case, combined with pytest-timeout thread method, the worker will crash out for a hang and
+# then it should get cleaned up by the controller through this fixture :fingers_crossed:
+@pytest.hookimpl(tryfirst=True)
+def pytest_handlecrashitem(crashitem, report, sched):
+    reset_tensix()
+
+
+def reset_tensix(tt_open_devices=None):
+    metal_env = copy.deepcopy(os.environ)
+    arch = metal_env.get("ARCH_NAME")
+    if arch != "grayskull" and arch != "wormhole_b0":
+        raise Exception(f"Unrecognized arch for tensix-reset: {arch}")
+
+    if tt_open_devices is None:
+        logger.info(f"Running reset with reset script: /opt/tt_metal_infra/scripts/ci/{arch}/reset.sh")
+        smi_reset_result = run_process_and_get_result(f"/opt/tt_metal_infra/scripts/ci/{arch}/reset.sh")
+    else:
+        tt_open_devices_str = ",".join([str(i) for i in tt_open_devices])
+        check_smi_metal = run_process_and_get_result("tt-smi-metal -h")
+        logger.info(f"Running reset for pci devices: {tt_open_devices_str}")
+        if check_smi_metal.returncode > 0:
+            logger.info(f"Test failed - resetting {arch} with tt-smi")
+            smi_reset_result = run_process_and_get_result(f"tt-smi -r {tt_open_devices_str}")
+        else:
+            smi_reset_result = run_process_and_get_result(f"tt-smi-metal -r {tt_open_devices_str}")
+    logger.info(f"tt-smi reset status: {smi_reset_result.returncode}")
+
+
+@pytest.hookimpl(tryfirst=True)
+def pytest_xdist_auto_num_workers(config):
+    logger.info("getting num of xdist workers")
```

Comment:
What's this?

Reply:
pytest fixture provided by `xdist` that lets us add logic to set # of workers used when you call `pytest -n auto`. This way no need to specify in pytest calls (can just use `auto`) and in the future can add logic using `GetNumAvailableDevices` or smthn similar to scale up.

Reply:
Will clean up the print